src/util/db/dbconnectionpool.cpp

```diff
@@ -0,0 +1,92 @@
+#include "util/db/dbconnectionpool.h"
+
+#include "util/memory.h"
+#include "util/logger.h"
+
+
+namespace mixxx {
+
+namespace {
+
+const mixxx::Logger kLogger("DbConnectionPool");
+
+} // anonymous namespace
+
+bool DbConnectionPool::createThreadLocalConnection() {
+    VERIFY_OR_DEBUG_ASSERT(!m_threadLocalConnections.hasLocalData()) {
+        DEBUG_ASSERT(m_threadLocalConnections.localData());
+        kLogger.critical()
+                << "Thread-local database connection already exists"
+                << *m_threadLocalConnections.localData();
+        return false; // abort
+    }
+    const int connectionIndex =
+            m_connectionCounter.fetchAndAddAcquire(1) + 1;
+    const QString indexedConnectionName =
+            QString("%1-%2").arg(
+                    m_prototypeConnection.name(),
+                    QString::number(connectionIndex));
+    auto pConnection = std::make_unique<DbConnection>(m_prototypeConnection, indexedConnectionName);
+    if (!pConnection->open()) {
+        kLogger.critical()
+                << "Failed to open thread-local database connection"
+                << *pConnection;
+        return false; // abort
+    }
+    m_threadLocalConnections.setLocalData(pConnection.get()); // transfer ownership
+    pConnection.release(); // release ownership
+    DEBUG_ASSERT(m_threadLocalConnections.hasLocalData());
+    DEBUG_ASSERT(m_threadLocalConnections.localData());
+    kLogger.info()
+            << "Cloned thread-local database connection"
+            << *m_threadLocalConnections.localData();
+    return true;
+}
+
+void DbConnectionPool::destroyThreadLocalConnection() {
+    m_threadLocalConnections.setLocalData(nullptr);
+}
+
+QSqlDatabase DbConnectionPool::threadLocalConnection() const {
```

Comment:
Shared users of the connection within the same thread do not need to know were the lifetime of the thread-local connection is actually managed. The DbConnection::LocalThreadScoped reference should not be passed around.

Instead users needing a connection temporarily for executing queries may ask the pool for one. But they must be prepared to get none, if the lifetime of the connection has already ended.

If we wrap DbConnectionPool::ThreadLocalScoped into a smart_ptr I see the danger that those references get passed between threads which is strictly forbidden and would lead into lifetime management hell. DbConnectionPool::ThreadLocalScoped should not be allocated dynamically, but as a local variable in the run() function of a thread.